File: Data Structures & Algorithms/search-2d-matrix/submission-0.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        boolean exist = false;

        // find row
        int r = matrix.length - 1;
        int l = 0;
        int targetRow = 0;

        while(r >= l){
            int mid = l + (r - l) / 2;
            if(matrix[mid][0] <= target && target <= matrix[mid][0]){
                targetRow = mid;
                break;
            } else if(matrix[mid][0] > target) r = mid - 1;
            else l = mid + 1;
        }

        // find target
        l = 0;
        r = matrix[targetRow].length - 1;

        while (r >= l){
            int mid = l + (r - l) / 2;
            int curNum = matrix[targetRow][mid];

            if(curNum > target) r = mid - 1;
            else if(curNum < target) l = mid + 1;
            else return true;
        }

        return exist;
    }

}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1 
        while r > l:
            mid = l + (r - l) // 2
            if nums[mid] > nums[r]:
                l = mid + 1
            else:
                r = mid

        pivot = l
        def bs(l, r):
            while r >= l:
                mid = l + (r - l) // 2

                if nums[mid] > target:
                    r = mid - 1
                elif nums[mid] < target:
                    l = mid + 1
                else:
                    return mid
            return -1

        p1 = bs(0, pivot - 1)
        return bs(pivot, len(nums) - 1) if p1 == -1 else p1
        